firedrake/rf_eigensolver.py

```diff
@@ -0,0 +1,661 @@
+
+        self.es.setDimensions(nev=self.n_evals, ncv=self.ncv, mpd=self.mpd)
+        self.es.setOperators(A, M)
+        with self.options_manager.inserted_options():
+            self.es.solve()
+        nconv = self.es.getConverged()
+        if nconv == 0:
+            raise ConvergenceError("Did not converge any eigenvalues.")
+        return nconv
+
+    def check_es_convergence(self):
+        r"""Check the convergence of the eigenvalue problem."""
+        r = self.es.getConvergedReason()
+        try:
+            reason = SLEPc.EPS.ConvergedReasons[r]
+        except KeyError:
+            reason = ("unknown reason (petsc4py enum incomplete?), "
+                      "try with -eps_converged_reason")
+        if r < 0:
+            raise ConvergenceError(
+                f"Eigenproblem failed to converge after {self.es.getIterationNumber()} iterations.\n"
+                f"Reason:\n{reason}"
+            )
+
+    def eigenvalue(self, i: int):
+        r"""Return the i-th eigenvalue of the solved problem."""
+        return self.es.getEigenvalue(i)
+
+    def eigenfunction(self, i: int):
+        r"""Return the i-th eigenfunction of the solved problem.
+
+        Returns
+        -------
+        (Function, Function)
+            The real and imaginary parts of the eigenfunction.
+        """
+        if self._problem.restrict:
+            eigenmodes_real = Function(self._problem.restricted_space)
+            eigenmodes_imag = Function(self._problem.restricted_space)
+        else:
+            eigenmodes_real = Function(self._problem.output_space)
+            eigenmodes_imag = Function(self._problem.output_space)
+        with eigenmodes_real.dat.vec_wo as vr:
+            with eigenmodes_imag.dat.vec_wo as vi:
+                self.es.getEigenvector(i, vr, vi)
+        if self._problem.restrict:
+            eigenmodes_real = Function(self._problem.output_space).interpolate(eigenmodes_real)
+            eigenmodes_imag = Function(self._problem.output_space).interpolate(eigenmodes_imag)
+        return eigenmodes_real, eigenmodes_imag
+
+
+def new_restricted_control_variable(reduced_functional: ReducedFunctional, function_space: WithGeometry, dual: bool=False):
+    """Return new variables suitable for storing a control value or its dual
+        by interpolating into the space over which the 'ReducedFunctional' is
+        defined.
+
+    Parameters
+    ----------
+        reduced_functional: The `ReducedFunctional` whose
+        controls are to be copied.
+        function_space: Function space to which we restrict variables to.
+        dual: whether to return a dual type. If False then a primal type is returned.
+
+    Returns
+    -------
+        tuple[OverloadedType]: New variables suitable for storing a control value.
+    """
+    return tuple(
+        Function(function_space).interpolate(control.control)._ad_init_zero(dual=dual)
+        for control in reduced_functional.controls
+    )
```

Comment:
I don't think there's any need for the interpolation if we're just asking for a zero-initialised object with `_ad_init_zero`, and `Function` is already zero-initialised
```suggestion
    return tuple(
        Function(function_space.dual() if dual else function_space)
        for control in reduced_functional.controls
    )
```